README.md
# pyMCS2
simple python wrapper for smartACT MCS2 ASCII commands


RELEASE_NOTES.md
29-Aug-2023 N. Balakrishnan
Added Log folder
Updated log file formatting
Added CBXFEL checkout document notes to log file

17-Aug-2023 N. Balakrishnan
Initial Version


testANL.py
from SmarAct2 import SmarActController2
#import cbxfelTestLogConst
from datetime import datetime
import logging

tester = input("Tester Name and Initials: \n")

print("Select test schedule phase:")
print("1:Pre-bakeout")
print("2:Pre-shipment")
print("3:Pre-installation")
print("4:Post-installation")
testSch = input("Select from 1 to 4: \n")

print("Select test location:")
print("1:ANL Lab/Workshop")
print("2:SLAC Lab/Workshop")
print("3:Undulator Hall")
testLoc = input("Select from 1 to 3: \n")

print("Select station ID:")
print("1:C")
print("2:D")
statID = input("Select from 1 or 2: \n")

if statID == "1":
  print("Select component ID:")
  print("1:X24")
  compID = input("Select from 1: \n")
    
if statID == "2":
  print("Select component ID:")
  print("1:X23")
  print("2:X23G")
  print("3:X24E")
  compID = input("Select from 1 to 3: \n")

print("Select axis ID:")
print("1:X")
print("2:Y")
axisID = input("Select from 1 or 2: \n")

stageSN = input("Enter serial number etched at the side of the stage: \n")

chan = input("Enter controller channel: \n")

now = datetime.now()

IP = input("Enter IP: \n")
units= 1000000000
 
date_time = now.strftime("%m%d%Y_%H_%M")
file_name = 'logs/{}_{}.log'.format(stageSN, date_time)

#start the logger
logging.basicConfig(filename=file_name, filemode='w', format='%(name)s - %(levelname)s - %(message)s', level= logging.INFO)
logging.info("BEGINNING TEST")
logging.info("Test location: {}".format(testLoca))
logging.info("Station ID: {}".format(stationID))
logging.info("Axis ID: {}".format(axisID))
logging.info("Stage serial number: {}".format(stageSN))
logging.info("Date: {}, Time: {}".format(now.strftime("%m/%d/%Y"), now.strftime("%H:%M:%S")))
logging.info("\n")

smart = SmarActController2(IP)
smart.create()
smart.connect()
print("Serial Number: {}".format(smart.get_serial_num()))
print("Positioner type Name: {}".format(smart.get_positioner_type_name(chan)))
print("Positioner reference type: {}".format(smart.get_positioner_reference_type(chan)))
smart.set_test_constraints(chan,units)
print("Max closed loop frequency (Hz): {}".format(smart.get_max_closed_loop_freq(chan)))
print("Positioner velocity (mm/sec): {}".format(smart.get_velocity(chan,units)))
print("Positioner acceleration (mm/sec2): {}".format(smart.get_acceleration(chan,units)))
print("Current position (mm): {}".format(smart.get_position(chan,units)))

dirnText = input("Move direction [Relative] (pos/neg): \n")
tweak = input("Move the axis by (mm) [Relative]: \n")

if dirnText == "pos":
 dirn = 1
if dirnText == "neg":
 dirn = 0
print(f'direction text = {dirnText} and dirn = {dirn}')

smart.move_relative(tweak,chan,units,dirn)
print("Current position (mm): {}".format(smart.get_position(chan,units)))
moveConfirmation =input("Confirm visually that the stage has moved (yes/no): \n")

logging.info("Confirm visually that the stage has moved (yes/no): {}".format(moveConfirmation))

moveDirection =input("Confirm direction of motion (pos/neg): \n")
logging.info("Confirm direction of motion (pos/neg): {}".format(moveDirection))

dirnText = input("Move direction [Relative] (pos/neg ): \n")
tweak = input("Move the axis by (mm) [Relative]: \n")
if dirnText == "pos":
 dirn = 1
if dirnText == "neg":
 dirn = 0
smart.move_relative(tweak,chan,units,dirn)

print("Current position (mm): {}".format(smart.get_position(chan,units)))

moveConfirmation =input("Confirm visually that the stage has moved (yes/no): \n")
logging.info("Confirm visually that the stage has moved (yes/no): {}".format(moveConfirmation))

moveDirection =input("Confirm direction of motion (pos/neg): \n")
logging.info("Confirm direction of motion (pos/neg): {}".format(moveDirection))

dirConfirmation =input("Are physical positive and negative directions of motion consistent with definitions (yes/no): \n")
logging.info("Are physical positive and negative directions of motion consistent with definitions (yes/no): {}".format(dirConfirmation))

findReference =input("Find Reference ? (yes/no): \n")
if findReference == "yes":
 smart.find_reference(chan)
else:
 print("Skip finding reference")

print("Current position (mm): {}".format(smart.get_position(chan,units)))

tweak = input("Move the axis to (mm) [Absolute]: \n")
smart.move_absolute(chan,tweak,units)
print("Current position (mm): {}".format(smart.get_position(chan,units)))

tweak = input("Move the axis to (mm) [Absolute]: \n")
smart.move_absolute(chan,tweak,units)
print("Current position (mm): {}".format(smart.get_position(chan,units)))

limit = input("Find Limit (pos/neg): \n")
smart.find_limit(chan,limit,units)
limit1 = smart.get_position(chan,units)
print("Current end of travel position (mm): {}".format(limit1))
logging.info("Current end of travel position (mm): {}".format(limit1))
limit = input("Find Limit (pos/neg): \n")
smart.find_limit(chan,limit,units)
limit2 = smart.get_position(chan,units)
print("Current end of travel position (mm): {}".format(limit2))
logging.info("Current end of travel position (mm): {}".format(limit2))
totalTravel = abs(limit1) + abs(limit2)
print("Total Range (mm): {}".format(totalTravel))
logging.info("Total Range (mm): {}".format(totalTravel))
midTravel = totalTravel/2
if limit == "pos":
 tweak = midTravel
 dirn = 0
if limit == "neg":
 tweak = midTravel
 dirn = 1
print("Moving Axis to mid point by (mm): {}".format(tweak))
logging.info("Moving Axis to mid point by (mm): {}".format(tweak))
smart.move_relative(tweak,chan,units,dirn)
print("Current position (mm): {}".format(smart.get_position(chan,units)))

smart.disconnect()



SmarAct2.py
import socket
import time
import logging
from mcs2Status import controllerStatus

class SmarActController2:
    def __init__(self, dev_ip) -> None:
        self.ip = dev_ip
        self.client = None
        pass
    
    def create(self):
        try:
            logging.info("Creating socket for {} port".format(self.ip))
            self.client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            timeout_seconds = 30
            self.client.settimeout(timeout_seconds)
            logging.info("Socket created.")
        except Exception as e:
            print("Failed to create socket")
            logging.critical("Failed to create socket: " + str(e))
            raise Exception("Failed to create socket: " + str(e))
        return None

    def connect(self):
        try:
            logging.info("Connecting to {} port".format(self.ip))
            self.client.connect((self.ip,55551))
            logging.info("Connected to {}\n".format(self.ip))
        except Exception as e:
            print("Unable to connect")
            logging.critical("Unable to connect: " + str(e))
            raise Exception("Unable to connect: " + str(e))
        return None

    def disconnect(self):
        self.client.close()
        logging.info("Disconnected to {}".format(self.ip))
    
    def send_no_response(self,command):
        self.client.send(bytes(command+'\r\n','ascii'))

    def send(self,command):
        self.client.send(bytes(command+'\r\n','ascii'))
        try:
            msg = self.client.recv(1024).decode('ascii')
        except Exception as e:
            print("No Response. Timeout")
            logging.critical("No Response. Timeout: " + str(e))
            raise Exception("No Response. Timeout: " + str(e))
        msg = msg.strip()
        return msg
    
    def get_serial_num(self):
        msg = self.send(":DEV:SNUM?")
        logging.info("Get serial number: {}". format(msg))
        return msg

    def get_sensor_type(self):
        msg = self.send(":DEV:PTYP?")
        logging.info("Get sensor type: {}". format(msg))
        return msg

    def get_positioner_type_name(self,channel:int):
        msg = self.send(":CHAN{}:PTYP:NAME?".format(channel))
        logging.info("Get positioner type name: {}". format(msg))
        return msg

    def get_position(self,channel:int,units: int):
        msg = self.send(":CHAN{}:POS:CURR?".format(channel))
        logging.info("Current position (units): {}". format(msg))
        posn = int(msg)/units
        logging.info("Current position (mm): {}\n". format(posn))
        return posn

    def get_max_closed_loop_freq(self,channel:int):
        msg = self.send(":CHAN{}:MCLF?".format(channel))
        logging.info("Max Closed Loop Frequency (Hz): {}". format(msg))
        return msg

    def get_positioner_reference_type(self,channel:int):
        msg = self.send(":CHAN{}:TUN:RTYP?".format(channel))
        if int(msg) == 0:
            logging.info("Positoner reference type: None\n")
        elif int(msg) == 1:
         logging.info("Positoner reference type: Hard Stop\n")
        elif int(msg) == 2:
         logging.info("Positoner reference type: Single Coded\n")
        elif int(msg) == 3:
         logging.info("Positoner reference type: Distance Coded\n")
        else: 
            logging.info("Error in querying positioner reference\n")
        return msg

    def get_velocity(self,channel:int,units: int):
        msg = self.send(":CHAN{}:VEL?".format(channel))
        logging.info("Positoner velocity (units/sec): {}". format(msg))
        velo = int(msg)/units
        logging.info("Positoner velocity (mm/sec): {}". format(velo))
        return (velo)

    def get_acceleration(self,channel:int,units: int):
        msg = self.send(":CHAN{}:ACC?".format(channel))
        logging.info("Positoner acceleration (units/sec2): {}". format(msg))
        accl = int(msg)/units
        logging.info("Positoner acceleration (mm/sec2): {}\n". format(accl))
        return (accl)

    def move_absolute(self,channel:int,position: int, units: int):
        timeout = 240
        time_elapsed = []
        status = self.get_status("{}".format(channel))
        mcs2 = controllerStatus(channel)
        if not mcs2.status_is_enabled(status):
           print("Amplifier not enabled. Enabling now")
           self.send_no_response(":CHAN{}:AMPL:ENAB 1".format(channel))
        timeout_elapsed = False
        self.send_no_response(":CHAN{}:MMODE 0".format(channel))
        logging.info("Move to position: {}". format(position))
        position = int(position)*units
        start_time = time.time()
        self.send_no_response(":MOVE{} {}".format(channel,position))
        time.sleep(0.1)
        status = self.get_status("{}".format(channel))
        while not timeout_elapsed :
            time.sleep(0.1)
            status = self.get_status("{}".format(channel))
            if not mcs2.status_is_moving(status):
               print("Move complete.")
               logging.info("Axis move complete")
               timeout_elapsed = True
            if (time.time()-start_time) > timeout:
                logging.info("Channel {} polling timeout elapsed ({}s)". format(channel, timeout))
                timeout_elapsed = True
                print("Timeout Elapsed")
        status = self.get_status("{}".format(channel))
        if mcs2.status_end_stop_reached(status):
           logging.info("Hard stop reached")

    def find_limit(self,channel:int,direction, units: int):
        timeout = 240
        time_elapsed = []
        status = self.get_status("{}".format(channel))
        mcs2 = controllerStatus(channel)
        if not mcs2.status_is_enabled(status):
           print("Amplifier not enabled. Enabling now")
           self.send_no_response(":CHAN{}:AMPL:ENAB 1".format(channel))
        timeout_elapsed = False
        self.send_no_response(":CHAN{}:MMODE 0".format(channel))
        logging.info("Move to find {} limit". format(direction))
        if direction == "pos":
         position = 30*units
        if direction == "neg":
         position = -30*units
        start_time = time.time()
        self.send_no_response(":MOVE{} {}".format(channel,position))
        time.sleep(0.1)
        status = self.get_status("{}".format(channel))
        while not timeout_elapsed :
            time.sleep(0.1)
            status = self.get_status("{}".format(channel))
            if not mcs2.status_is_moving(status):
               print("Move complete.")
               logging.info("Axis move complete")
               timeout_elapsed = True
            if (time.time()-start_time) > timeout:
                logging.info("Channel {} polling timeout elapsed ({}s)". format(channel, timeout))
                timeout_elapsed = True
                print("Timeout Elapsed")
        status = self.get_status("{}".format(channel))
        if mcs2.status_end_stop_reached(status):
           print("Reached hard stop.")
           logging.info("Hard stop reached")
 
    def move_relative(self,tweak: int,channel:int,units: int,dirn):
        timeout = 300
        time_elapsed = []
        status = self.get_status("{}".format(channel))
        mcs2 = controllerStatus(channel)
        if not mcs2.status_is_enabled(status):
           print("Amplifier not enabled. Enabling now")
           self.send_no_response(":CHAN{}:AMPL:ENAB 1".format(channel))
        timeout_elapsed = False
        self.send_no_response(":CHAN{}:MMODE 1".format(channel))
        start_time = time.time()
        if dirn == 1:
           logging.info("Move in positive direction in mm by {}". format(tweak))
           posn = int(tweak)*units
        else:
           logging.info("Move in negative direction in mm by {}". format(tweak))
           posn = int(tweak)*units*(-1)
           
        self.send_no_response(":MOVE{} {}".format(channel,posn))
        time.sleep(0.1)
        status = self.get_status("{}".format(channel))
        while not timeout_elapsed :
            time.sleep(0.1)
            status = self.get_status("{}".format(channel))
            if not mcs2.status_is_moving(status):
               print("Move complete.")
               logging.info("Axis move complete")
               timeout_elapsed = True
            if (time.time()-start_time) > timeout:
                logging.info("Channel {} polling timeout elapsed ({}s)". format(channel, timeout))
                timeout_elapsed = True
                print("Timeout Elapsed")
        status = self.get_status("{}".format(channel))
        if mcs2.status_end_stop_reached(status):
           print("Reached hard stop.")
           logging.info("Hard stop reached")

    def find_reference(self,channel:int):
        timeout = 240
        time_elapsed = []
        status = self.get_status("{}".format(channel))
        mcs2 = controllerStatus(channel)
        if not mcs2.status_is_enabled(status):
           print("Amplifier not enabled. Enabling now")
           self.send_no_response(":CHAN{}:AMPL:ENAB 1".format(channel))
        if mcs2.status_is_referenced(status):
           findReference =input("Axis referenced. Find Reference again? (y/n): \n")
        else:
           findReference = "y"
        if findReference == "y":
              self.send_no_response(":CHAN{}:REF:OPT 0".format(channel))
              logging.info("Setting ref option to stop at reference mark and set value to 0.")
              self.send_no_response(":REF{}".format(channel))
              timeout_elapsed = False
              start_time = time.time()
              status = self.get_status("{}".format(channel))
              while not timeout_elapsed :
                  time.sleep(0.1)
                  status = self.get_status("{}".format(channel))
                  if not mcs2.status_is_referencing(status):
                     print("Referencing complete.")
                     logging.info("Axis referencing complete")
                     timeout_elapsed = True
                  if (time.time()-start_time) > timeout:
                      logging.info("Axis {} polling timeout elapsed ({}s)". format(channel, timeout))
                      timeout_elapsed = True
                      print("Timeout Elapsed")
              print("Referencing complete")
        else:
              print("Skip finding reference")
        time.sleep(0.1)
        status = self.get_status("{}".format(channel))
        if mcs2.status_end_stop_reached(status):
           print("Reached hard stop.")
           logging.info("Hard stop reached")
        if mcs2.status_is_referenced(status):
           print("Axis referenced.")
           logging.info("Axis referenced")

    def set_test_constraints(self,channel:int,units: int):
        velo = 0.2 * units
        accl = 0.1 * units
        print("\nSetting test constraints.")
        print("Maximum closed loop frequency = 100 Hz.")
        print("Closed loop move velocity = 0.2 mm/s.")
        print("Maximum axis acceleration = 0.1 mm/s^2.")
        logging.info("Setting test constraints.")
        logging.info("Maximum closed loop frequency = 100 Hz.")
        logging.info("Closed loop move velocity = 0.2 mm/s.")
        logging.info("Maximum axis acceleration = 0.1 mm/s^2.")
        self.send_no_response(":CHAN{}:MCLF 100".format(channel))
        self.send_no_response(":CHAN{}:VEL {}".format(channel,velo))
        self.send_no_response(":CHAN{}:ACC {}".format(channel,accl))
        logging.info("Test Constraints set.\n")
        print("Test Constraints set.\n")

    def get_test_constraints(self,channel:int,units: int):
        print("Reading test constraints from controller.")
        logging.info("Reading test constraints.")
        print("Maximum closed loop frequency (Hz): {}".format(self.get_max_closed_loop_freq(channel)))
        print("Closed loop move velocity (mm/s): {}".format(self.get_velocity(channel,units)))
        print("Maximum axis acceleration (mm/s^2): {}\n".format(self.get_acceleration(channel,units)))

    def get_status(self, channel:int):
        msg = self.send(":CHAN{}:STAT?".format(channel))
        mcs2 = controllerStatus(channel)
        return msg




mcs2Status.py
import logging

class controllerStatus:
    def __init__(self, channel) -> None:
        self.chan = channel
        pass
    
    def status_is_referencing(self, status):
        status = int(status)
        is_referencing = int ('8',16)
        result = (status & is_referencing)
        if (result) == (is_referencing):
            return 1
        else:
            return 0

    def status_is_moving(self, status):
        status = int(status)
        move_active = int ('1',16)
        result = (status & move_active)
        if (result) == (move_active):
            return 1
        else:
            return 0
    
    def status_is_enabled(self, status):
        status = int(status)
        is_enabled = int ('40000',16)
        result = (status & is_enabled)
        if int(result) == is_enabled:
            logging.info("Axis enabled")
            return 1
        else:
            logging.info("Axis disabled")
            return 0
    
    def status_end_stop_reached(self, status):
        status = int(status)
        end_stop = int ('100',16)
        result = (status & end_stop)
        if (result) == (end_stop):
            return 1
        else:
            return 0

    def status_is_referenced(self, status):
        status = int(status)
        is_referenced = int ('80',16)
        result = (status & is_referenced)
        if (result) == (is_referenced):
            return 1
        else:
            return 0




cbxfelCheckoutLog.py
import logging
from datetime import datetime

test_schedule = {
    "1":"Pre Bakeout",
    "2":"Pre Shipment",
    "3":"Pre Installation",
    "4":"Post Intallation"
}

test_loca =  {
    "1":"ANL Lab/Workshop",
    "2":"SLAC Lab/Workshop",
    "3":"Undulator Hall"
}

stat_ID = {
    "1":"C",
    "2":"D"
}

component_ID_C = {
    "1":"X23", 
    "2":"X23G",
    "3":"X23E"
}

component_ID_D = {
    "1":"X24"
}

axis_ID = {
    "1":"X",
    "2":"Y"
}

class cbxfelLog:
   def __init__(self):
       cbxfelLog.tester_info()

       print("\nSelect test location:")
       print("1:ANL Lab/Workshop")
       print("2:SLAC Lab/Workshop")
       print("3:Undulator Hall")
       testLoca = input("Enter 1 or 2 or 3: \n")
       cbxfelLog.test_location(testLoca)

       logging.info("Item 1:")
       print("\nItem 1:")
       print("Select test schedule phase:")
       print("1:Pre-bakeout")
       print("2:Pre-shipment")
       print("3:Pre-installation")
       print("4:Post-installation")
       testSch = input("Enter 1 or 2 or 3 or 4: \n")
       cbxfelLog.test_schedule_phase(testSch)

       logging.info("Item 2:")
       print("\nItem 2:")
       print("Select station ID:")
       print("1:C")
       print("2:D")
       statID = input("Enter 1 or 2: \n")
       cbxfelLog.station_ID(statID)

       logging.info("Item 3:")
       print("\nItem 3:")
       cbxfelLog.component_ID(statID)

       logging.info("Item 4:")
       print("\nItem 4:")
       print("Select axis ID:")
       print("1:X")
       print("2:Y")
       axisID = input("Enter 1 or 2: \n")
       cbxfelLog.axis_xy_ID(axisID)
       
       logging.info("Item 5:")
       print("\nItem 5:")
       print("Confirm the controller channel that this axis is assigned to in SLAC-I-120-174 matches the physical connection:")
       msg = input("y or n: \n")
       logging.info("Confirm the controller channel that this axis is assigned to in SLAC-I-120-174 matches the physical connection:")
       logging.info("y or n: {}\n".format(msg))

   def tester_info():
      tester = input("\nTester Name and Initials: \n")
      logging.info("Tester Name and Initials: {}".format(tester))
      now = datetime.now()
      logging.info("Date: {}, Time: {}".format(now.strftime("%m/%d/%Y"), now.strftime("%H:%M:%S")))
          
   def test_schedule_phase(msg):
      try:
         print("Test Schedule Phase: ",test_schedule[msg])
         logging.info("Test Schedule Phase: {}\n".format(test_schedule[msg]))
      except KeyError:
         logging.error("Invalid Test Schedule Phase: {}".format(msg))
         logging.error("Enter 1 or 2 or 3 or 4")
         print("Invalid Test Schedule Phase: ",msg)
         testSch = input("Enter 1 or 2 or 3 or 4: \n")
         cbxfelLog.test_schedule_phase(testSch)

   def test_location(msg):
      try:
         print("Test Location: ",test_loca[msg])
         logging.info("Test Location: {} \n".format(test_loca[msg]))
      except KeyError:
         logging.error("Invalid Test Location: {}".format(msg))
         logging.error("Enter 1 or 2 or 3")
         print("Invalid Test Location: ",msg)
         testLoca = input("Enter 1 or 2 or 3: \n")
         cbxfelLog.test_location(testLoca)

   def station_ID(msg):
      try:
         print("Station ID: ",stat_ID[msg])
         logging.info("Station ID: {}\n".format(stat_ID[msg]))
      except KeyError:
         logging.error("Invalid station ID: {}".format(msg))
         logging.error("Enter 1 or 2")
         print("Invalid station ID: ",msg)
         statID = input("Enter 1 or 2: \n")
         cbxfelLog.station_ID(statID)

   def component_ID(msg):
      if msg == "2":
        cID = component_ID_D['1']
        print("\nComponent ID: ",cID)
        logging.info("Component ID: {}\n".format(cID))
      else:
        print("\nSelect component ID:")
        print("1:X23")
        print("2:X23G")
        print("3:X23E")
        compID = input("Select from 1 to 3: \n")
        cID = component_ID_C[compID]
        try:
           print("Component ID: ",cID)
           logging.info("Component ID: {}\n".format(cID))
        except KeyError:
           logging.error("Invalid component ID: {}".format(compID))
           logging.error("Enter 1 or 2 or 3")
           print("Invalid component ID: ",compID)

   def axis_xy_ID(msg):
      try:
         print("Axis ID: ",axis_ID[msg])
         logging.info("Axis ID: {}\n".format(axis_ID[msg]))
      except KeyError:
         logging.error("Invalid axis ID: {}".format(msg))
         logging.error("Enter 1 or 2")
         print("Invalid axis ID: ",msg)
         axisID = input("Enter 1 or 2: \n")
         cbxfelLog.axis_xy_ID(axisID)



testSmartActEuclid.py
from SmarAct2 import SmarActController2
from cbxfelCheckoutLog import cbxfelLog
from datetime import datetime
import logging
import json


stageSN = input("\nEnter serial number etched at the side of the stage. This will be the filename.\n")

now = datetime.now()
date_time = now.strftime("%m%d%Y_%H_%M")
file_name = 'logs/{}_{}.log'.format(stageSN, date_time)

#start the logger
logging.basicConfig(filename=file_name, filemode='w', format='%(name)s - %(levelname)s - %(message)s', level= logging.INFO)
logging.info("BEGINNING TEST")
logging.info("Stage serial number: {}".format(stageSN))

cbxfelLog()

IP = "192.168.1.10"
units= 1000000000

print("\nIP: {}".format(IP))
logging.info("IP: {}\n".format(IP))

logging.info("Item 6:")
print("\nItem 6:")
chan = input("Enter controller channel (0/1/2): \n")
logging.info("Enter controller channel (0/1/2): {}".format(chan))

print("\nConfirm that the SmarAct Motion Control Software is configured to control the correct axis:")
msg = input("y or n: \n")
logging.info("Confirm that the SmarAct Motion Control Software is configured to control the correct axis:")
logging.info("y or n: {} \n".format(msg))

smart = SmarActController2(IP)
smart.create()
smart.connect()

print("\nSerial Number: {}".format(smart.get_serial_num()))
print("Positioner type Name: {}".format(smart.get_positioner_type_name(chan)))
print("Positioner reference type: {}".format(smart.get_positioner_reference_type(chan)))
smart.set_test_constraints(chan,units)
smart.get_test_constraints(chan,units)

logging.info("Item 7:")
print("\nItem 7:")
print("Confirm that the frequency, velocity, and acceleration parameters are consistent with the restrictions in Section 9.2 of this document.")
msg = input("y or n: \n")
logging.info("Confirm that the frequency, velocity, and acceleration parameters are consistent with the restrictions in Section 9.2 of this document.")
logging.info("y or n:{} \n".format(msg))

logging.info("Item 8:")
print("\nItem 8:")
print("Inital actuator position (mm): {}".format(smart.get_position(chan,units)))


logging.info("Item 9:")
print("\nItem 9:")
print("Command an incremental movement of 1.0 mm in positive direction")
logging.info("Command an incremental movement of 1.0 mm in positive direction")
msg = input("y or n: \n")
logging.info("y or n: {}\n".format(msg))

if msg == "y":
 smart.move_relative(1,chan,units,1)
 moveConfirmation =input("Confirm visually that the stage has moved by correct amount and in the correct direction (y/n): \n")
 logging.info("Confirm visually that the stage has moved by correct amount and in the correct direction (y/n): {}".format(moveConfirmation))
 
 moveDirection =input("Report direction of motion (p/n): \n")
 logging.info("Report direction of motion (p/n): {}".format(moveDirection))
else:
 print("Response was either n or skipped")
 logging.info("Response was either n or skipped")
 print("Cannot confirm visually that the stage has moved. \n")
 logging.info("Cannot confirm visually that the stage has moved. \n")
 print("Cannot report direction of motion. \n")
 logging.info("Cannot report direction of motion. \n")


print("Current position (mm): {}".format(smart.get_position(chan,units)))

logging.info("Item 10:")
print("\nItem 10:")
print("Command an incremental movement of 1.0 mm in negative direction")
logging.info("Command an incremental movement of 1.0 mm in negative direction")
msg = input("y or n: \n")
logging.info("y or n: {} \n".format(msg))

if msg == "y":
 smart.move_relative(1,chan,units,0)
 moveConfirmation =input("Confirm visually that the stage has moved by correct amount and in the correct direction (y/n): \n")
 logging.info("Confirm visually that the stage has moved by correct amount and in the correct direction (y/n): {}".format(moveConfirmation))
 moveDirection =input("Report direction of motion (p/n): \n")
 logging.info("Report direction of motion (p/n): {}".format(moveDirection))
else:
 print("Response was either n or skipped\n")
 logging.info("Response was either n or skipped\n")
 print("Cannot confirm visually that the stage has moved. \n")
 logging.info("Cannot confirm visually that the stage has moved. \n")
 print("Cannot report direction of motion. \n")
 logging.info("Cannot report direction of motion. \n")


print("Current position (mm): {}".format(smart.get_position(chan,units)))

logging.info("Item 11:")
print("\nItem 11:")
dirConfirmation =input("Are physical positive and negative directions of motion consistent with directions of motions defined in SLAC-I-120-005 (y/n): ")
logging.info("Are physical positive and negative directions of motion consistent with directions of motions defined in SLAC-I-120-005 (y/n): {}\n".format(dirConfirmation))

logging.info("Item 12:")
print("\nItem 12:")
logging.info("Axis will be homed in positive direction.\n" )
print("Axis will be homed in positive direction.\n" )

logging.info("Item 13:")
print("\nItem 13:")

findReference =input("Find Reference in default positive direction ? (y/n): \n")
logging.info("Find Reference in default positive direction ? (y/n): {}".format(findReference))
if findReference == "y":
 smart.find_reference(chan)
 print("Current position (mm): {}".format(smart.get_position(chan,units)))
 moveConfirmation =input("Confirm the stage has moved to correct end of travel and readback is 0. (y/n): \n")
 logging.info("Confirm the stage has moved to correct end of travel and readback is 0. (y/n): {}\n".format(moveConfirmation))
else:
 print("Skip finding reference")
 logging.info("Skip finding reference")
 print("Skip confirming the stage has moved to correct end of travel and readback is 0.\n")
 logging.info("Skip confirming the stage has moved to correct end of travel and readback is 0.\n")


logging.info("Item 14:")
print("\nItem 14:")

limit = input("Incrementally move the stage in positive end of the range of motion.(y/n): \n")
logging.info("Incrementally move the stage in positive end of the range of motion.(y/n): {}".format(limit))
if limit == "y":
 smart.find_limit(chan,"pos",units)
 limit1 = smart.get_position(chan,units)
 print("Current end of travel position (mm): {}".format(limit1))
 logging.info("Current end of travel position (mm): {}".format(limit1))
 skip = 0
else:
 print("Skipped moving to positive end of travel")
 logging.info("Skipped moving to positive end of travel")
 skip = 1

logging.info("\nItem 15:")
print("\nItem 15:")

limit = input("Incrementally move the stage in negative end of the range of motion.(y/n): \n")
logging.info("Incrementally move the stage in negative end of the range of motion.(y/n): {}".format(limit))
if limit == "y":
 smart.find_limit(chan,"neg",units)
 limit2 = smart.get_position(chan,units)
 print("Current end of travel position (mm): {}".format(limit2))
 logging.info("Current end of travel position (mm): {}".format(limit2))
 skip = 0
else:
 print("Skipped moving to negative end of travel")
 logging.info("Skipped moving to negative end of travel")
 skip = 1

logging.info("\nItem 16:")
print("\nItem 16:")

if skip != 1:
 msg = input("Incrementally move the stage to midpoint of the range of motion.(y/n): \n")
 logging.info("Incrementally move the stage to midpoint of the range of motion.(y/n): {}".format(msg))
 if msg == "y":
   totalTravel = abs(limit1) + abs(limit2)
   midTravel = totalTravel/2
   tweak = midTravel
   dirn = 1
   smart.move_relative(tweak,chan,units,dirn)
   print("Current position (mm): {}\n".format(smart.get_position(chan,units)))
   
 else:
   print("Skipped finding midpoint of travel range\n")
   logging.info("Skipped finding midpoint of travel range\n")

 logging.info("Item 17:")
 print("\nItem 17:")
 print("Difference in motor positions at limits to determine stage travel range (mm): {}".format(totalTravel))
 logging.info("Difference in motor positions at limits to determine stage travel range (mm): {}".format(totalTravel))

else:
 logging.info("Item 16 and 17:")
 print("\nItem 16 and 17:")
 print("As we did not move to both end of travel, we cannot find travel range or mid-point. END")  
 logging.info("As we did not move to both end of travel, we cannot find travel range or mid-point. END")  

smart.disconnect()


